File: sdk/go/vorpal.go
package main

import (
	"bytes"
	"fmt"
	"text/template"

	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/artifact/language"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

type ReleaseScriptTemplateArgs struct {
	Aarch64Darwin string
	Aarch64Linux  string
	BranchName    string
	X8664Darwin   string
	X8664Linux    string
}

const ReleaseScriptTemplate = `
git clone \
    --branch {{.BranchName}} \
    --depth 1 \
    git@github.com:ALT-F4-LLC/vorpal.git

pushd vorpal

git fetch --tags
git tag --delete nightly || true
git push origin :refs/tags/nightly || true
gh release delete --yes nightly || true

git tag nightly
git push --tags

gh release create \
    --notes "Nightly builds from main branch." \
    --prerelease \
    --title "nightly" \
    --verify-tag \
    nightly \
    /var/lib/vorpal/store/{{.Aarch64Darwin}}.tar.zst`

func build(context *config.ConfigContext) (*string, error) {
	protoc, err := artifact.Protoc(context)
	if err != nil {
		return nil, err
	}

	name := "vorpal"

	return language.NewRustBuilder(name).
		WithArtifacts([]*string{protoc}).
		WithBins([]string{name}).
		WithPackages([]string{
			"crates/agent",
			"crates/cli",
			"crates/registry",
			"crates/schema",
			"crates/sdk",
			"crates/store",
			"crates/worker",
		}).
		Build(context)
}

func buildProcess(context *config.ConfigContext) (*string, error) {
	vorpal, err := build(context)
	if err != nil {
		return nil, err
	}

	entrypoint := fmt.Sprintf("%s/bin/vorpal", artifact.GetEnvKey(vorpal))

	return artifact.NewArtifactProcessBuilder("vorpal-process", entrypoint).
		WithArguments([]string{
			"--registry",
			"http://localhost:50051",
			"start",
			"--port",
			"50051",
		}).
		WithArtifacts([]*string{vorpal}).
		Build(context)
}

func buildRelease(context *config.ConfigContext) (*string, error) {
	varAarch64Darwin, err := artifact.
		NewArtifactVariableBuilder("aarch64-darwin").
		WithRequire().
		Build(context)
	if err != nil {
		return nil, err
	}

	varAarch64Linux, err := artifact.
		NewArtifactVariableBuilder("aarch64-linux").
		WithRequire().
		Build(context)
	if err != nil {
		return nil, err
	}

	varBranchName, err := artifact.
		NewArtifactVariableBuilder("branch-name").
		WithRequire().
		Build(context)
	if err != nil {
		return nil, err
	}

	varX8664Darwin, err := artifact.
		NewArtifactVariableBuilder("x8664-darwin").
		WithRequire().
		Build(context)
	if err != nil {
		return nil, err
	}

	varX8664Linux, err := artifact.
		NewArtifactVariableBuilder("x8664-linux").
		WithRequire().
		Build(context)
	if err != nil {
		return nil, err
	}

	aarch64Darwin, err := context.FetchArtifact(*varAarch64Darwin)
	if err != nil {
		return nil, err
	}

	aarch64Linux, err := context.FetchArtifact(*varAarch64Linux)
	if err != nil {
		return nil, err
	}

	gh, err := artifact.Gh(context)
	if err != nil {
		return nil, err
	}

	x8664Darwin, err := context.FetchArtifact(*varX8664Darwin)
	if err != nil {
		return nil, err
	}

	x8664Linux, err := context.FetchArtifact(*varX8664Linux)
	if err != nil {
		return nil, err
	}

	artifacts := []*string{
		aarch64Darwin,
		aarch64Linux,
		gh,
		x8664Darwin,
		x8664Linux,
	}

	scriptTemplate, err := template.New("script").Parse(ReleaseScriptTemplate)
	if err != nil {
		return nil, err
	}

	var scriptBuffer bytes.Buffer

	scriptTemplateVars := ReleaseScriptTemplateArgs{
		Aarch64Darwin: *aarch64Darwin,
		Aarch64Linux:  *aarch64Linux,
		BranchName:    *varBranchName,
		X8664Darwin:   *x8664Darwin,
		X8664Linux:    *x8664Linux,
	}

	if err := scriptTemplate.Execute(&scriptBuffer, scriptTemplateVars); err != nil {
		return nil, err
	}

	return artifact.NewArtifactTaskBuilder("vorpal-release", scriptBuffer.String()).
		WithArtifacts(artifacts).
		Build(context)
}

func buildShell(context *config.ConfigContext) (*string, error) {
	gh, err := artifact.Gh(context)
	if err != nil {
		return nil, err
	}

	gobin, err := artifact.GoBin(context)
	if err != nil {
		return nil, err
	}

	goimports, err := artifact.Goimports(context)
	if err != nil {
		return nil, err
	}

	gopls, err := artifact.Gopls(context)
	if err != nil {
		return nil, err
	}

	grpcurl, err := artifact.Grpcurl(context)
	if err != nil {
		return nil, err
	}

	protoc, err := artifact.Protoc(context)
	if err != nil {
		return nil, err
	}

	protocGenGo, err := artifact.ProtocGenGo(context)
	if err != nil {
		return nil, err
	}

	protocGenGoGRPC, err := artifact.ProtocGenGoGRPC(context)
	if err != nil {
		return nil, err
	}

	staticcheck, err := artifact.Staticcheck(context)
	if err != nil {
		return nil, err
	}

	vorpalProcess, err := buildProcess(context)
	if err != nil {
		return nil, err
	}

	nginx, err := artifact.Nginx(context)
	if err != nil {
		return nil, err
	}

	return language.NewRustShellBuilder("vorpal-shell").
		WithArtifacts([]*string{
			gh,
			gobin,
			goimports,
			gopls,
			grpcurl,
			nginx,
			protoc,
			protocGenGo,
			protocGenGoGRPC,
			staticcheck,
			vorpalProcess,
		}).
		Build(context)
}

func buildTest(context *config.ConfigContext) (*string, error) {
	vorpal, err := build(context)
	if err != nil {
		return nil, err
	}

	script := fmt.Sprintf("\n%s/bin/vorpal --version", artifact.GetEnvKey(vorpal))

	return artifact.NewArtifactTaskBuilder("vorpal-test", script).
		WithArtifacts([]*string{vorpal}).
		Build(context)
}

File: sdk/typescript/src/vorpal.ts
import { ArtifactSystem } from "./api/artifact/artifact.js";
import {
  getEnvKey,
  Job,
  OciImage,
  Process,
  DevelopmentEnvironment,
  UserEnvironment,
} from "./artifact.js";
import { Bun } from "./artifact/bun.js";
import { Crane } from "./artifact/crane.js";
import { GoBin } from "./artifact/go.js";
import { Goimports } from "./artifact/goimports.js";
import { Gopls } from "./artifact/gopls.js";
import { Grpcurl } from "./artifact/grpcurl.js";
import { Rust } from "./artifact/language/rust.js";
import { NodeJS } from "./artifact/nodejs.js";
import { Pnpm } from "./artifact/pnpm.js";
import { Protoc } from "./artifact/protoc.js";
import { ProtocGenGo } from "./artifact/protoc_gen_go.js";
import { ProtocGenGoGrpc } from "./artifact/protoc_gen_go_grpc.js";
import { Staticcheck } from "./artifact/staticcheck.js";
import { ConfigContext } from "./context.js";

const SYSTEMS: ArtifactSystem[] = [
  ArtifactSystem.AARCH64_DARWIN,
  ArtifactSystem.AARCH64_LINUX,
  ArtifactSystem.X8664_DARWIN,
  ArtifactSystem.X8664_LINUX,
];

function getGoarch(system: ArtifactSystem): string {
  switch (system) {
    case ArtifactSystem.AARCH64_DARWIN:
    case ArtifactSystem.AARCH64_LINUX:
      return "arm64";
    case ArtifactSystem.X8664_DARWIN:
    case ArtifactSystem.X8664_LINUX:
      return "amd64";
    default:
      throw new Error(`unsupported system for GOARCH: ${system}`);
  }
}

function getGoos(system: ArtifactSystem): string {
  switch (system) {
    case ArtifactSystem.AARCH64_DARWIN:
    case ArtifactSystem.X8664_DARWIN:
      return "darwin";
    case ArtifactSystem.AARCH64_LINUX:
    case ArtifactSystem.X8664_LINUX:
      return "linux";
    default:
      throw new Error(`unsupported system for GOOS: ${system}`);
  }
}

async function buildVorpal(context: ConfigContext): Promise<string> {
  return new Rust("vorpal", SYSTEMS)
    .withBins(["vorpal"])
    .withIncludes(["cli", "sdk/rust"])
    .withPackages(["vorpal-cli", "vorpal-sdk"])
    .build(context);
}

async function buildVorpalContainerImage(
  context: ConfigContext,
): Promise<string> {
  const linuxVorpalSlim = await context.fetchArtifactAlias(
    "linux-vorpal-slim:latest",
  );
  const vorpal = await buildVorpal(context);

  const name = "vorpal-container-image";

  return new OciImage(name, linuxVorpalSlim)
    .withAliases([`${name}:latest`])
    .withArtifacts([vorpal])
    .build(context);
}

async function buildVorpalJob(context: ConfigContext): Promise<string> {
  const vorpal = await buildVorpal(context);
  const script = `${getEnvKey(vorpal)}/bin/vorpal --version`;

  return new Job("vorpal-job", script, SYSTEMS)
    .withArtifacts([vorpal])
    .build(context);
}

async function buildVorpalProcess(context: ConfigContext): Promise<string> {
  const vorpal = await buildVorpal(context);

  return new Process(
    "vorpal-process",
    `${getEnvKey(vorpal)}/bin/vorpal`,
    SYSTEMS,
  )
    .withArguments([
      "--registry",
      "https://localhost:50051",
      "services",
      "start",
      "--port",
      "50051",
    ])
    .withArtifacts([vorpal])
    .build(context);
}

async function buildVorpalShell(context: ConfigContext): Promise<string> {
  const bun = await new Bun().build(context);
  const crane = await new Crane().build(context);
  const go = await new GoBin().build(context);
  const goimports = await new Goimports().build(context);
  const gopls = await new Gopls().build(context);
  const grpcurl = await new Grpcurl().build(context);
  const nodejs = await new NodeJS().build(context);
  const pnpm = await new Pnpm().build(context);
  const protoc = await new Protoc().build(context);
  const protocGenGo = await new ProtocGenGo().build(context);
  const protocGenGoGrpc = await new ProtocGenGoGrpc().build(context);
  const staticcheck = await new Staticcheck().build(context);

  const goarch = getGoarch(context.getSystem());
  const goos = getGoos(context.getSystem());

  return new DevelopmentEnvironment("vorpal-shell", SYSTEMS)
    .withArtifacts([
      bun,
      crane,
      go,
      goimports,
      gopls,
      grpcurl,
      nodejs,
      pnpm,
      protoc,
      protocGenGo,
      protocGenGoGrpc,
      staticcheck,
    ])
    .withEnvironments([
      "CGO_ENABLED=0",
      `GOARCH=${goarch}`,
      `GOOS=${goos}`,
    ])
    .build(context);
}

async function buildVorpalUser(context: ConfigContext): Promise<string> {
  return new UserEnvironment("vorpal-user", SYSTEMS)
    .withArtifacts([])
    .withEnvironments(["PATH=$HOME/.vorpal/bin"])
    .withSymlinks([
      [
        "$HOME/Development/repository/github.com/ALT-F4-LLC/vorpal.git/main/target/debug/vorpal",
        "$HOME/.vorpal/bin/vorpal",
      ],
    ])
    .build(context);
}

async function main(): Promise<void> {
  const context = ConfigContext.create();

  switch (context.getArtifactName()) {
    case "vorpal":
      await buildVorpal(context);
      break;
    case "vorpal-container-image":
      await buildVorpalContainerImage(context);
      break;
    case "vorpal-job":
      await buildVorpalJob(context);
      break;
    case "vorpal-process":
      await buildVorpalProcess(context);
      break;
    case "vorpal-shell":
      await buildVorpalShell(context);
      break;
    case "vorpal-user":
      await buildVorpalUser(context);
      break;
    default:
      break;
  }

  await context.run();
}

main();
